molviewspec/molviewspec/builder.py

```diff
@@ -673,55 +715,321 @@ def transparency(self, *, transparency: float = 0.8) -> Representation:
+        params = make_params(MeshParams, {"kind": "mesh", **locals()})
+        node = Node(kind="primitive", params=params)
         self._add_child(node)
         return self
 
-    def line(
+    # TODO mesh_from_uri, mesh_from_source
+
+    # TODO need ellipsis support
+    def circle(
         self,
         *,
-        position1: tuple[float, float, float],
-        position2: tuple[float, float, float],
+        center: PositionT,
         radius: float,
-        color: ColorT,
-        label: str | None = None,
+        segments: int | None = None,
+        theta_start: float = 0,
+        theta_length: float = math.pi * 2,
+        rotation: Mat3[float] | None = None,
+    ) -> Primitives:
+        """
+        Add a circle.
+        :param center: Center coordinates of this item
+        :param radius: Radius of circle
+        :param segments: Number of segments used to approximate a circle.
+        :param theta_start: Start point position (relevant when this is an arc)
+        :param theta_length: Values < PI*2 will result in an arc
+        :param rotation: Optional: Control orientation of this item
+        :return: the corresponding geometric primitive builder, allowing further customization
+        """
+        params = make_params(CircleParams, {"kind": "circle", **locals()})
+        node = Node(kind="primitive", params=params)
+        self._add_child(node)
+        return self
+
+    def line(
+        self,
+        *,
+        start: PositionT,
+        end: PositionT,
+        thickness: float | None = 0.05,
+        dash_start: float | None = None,
+        dash_length: float | None = None,
+        gap_length: float | None = None,
+        color: ColorT | None = None,
         tooltip: str | None = None,
-    ) -> GenericVisuals:
-        """
-        Draw a line.
-        :param position1: start of line
-        :param position2: end of line
-        :param radius: width of line segment
-        :param color: color of line, either SVG color name or RGB hex code
-        :param label: optional text label
-        :param tooltip: optional tooltip to show upon hover
-        :return:
-        """
-        params = make_params(LineParams, locals())
-        node = Node(kind="line", params=params)
+    ) -> Primitives:
+        """
+        Defines a line, connecting a start and an end point.
+        :param start: origin coordinates
+        :param end: destination coordinates
+        :param thickness: thickness of this line
+        :param dash_start: offset along this line until the 1st dash is drawn
+        :param dash_length: length of each dash
+        :param gap_length: length of each gap that will follow each completed dash
+        :param color: color of the line
+        :param tooltip: tooltip to show when hovering the line
+        :return: this builder
+        """
+        params = make_params(LineParams, {"kind": "line", **locals()})
+        node = Node(kind="primitive", params=params)
```

Comment:
@midlik @JonStargaryen @aliaksei-chareshneu  I've changed the implementation to only have 1 `primitive` node kind and the type of the primitive being part of the `props` instead. This also much more nicely supports the "JSON data" shape